Enable new schema changer

fqazi opened the pull request.

ajwerner committed:
schemachanger: enable declarative schema changer by default

Release note: None

fqazi committed:
schemachanger: fix drop view errors with cross DB refs

Previously, the drop view errors generated by the declarative
schema changer did not have fully resolved names when cross
database references were involved. To be consistent with the
legacy schema changer we should add fully resolved names
when the parent databases differ. To address this, this
patch will add fully resolved names on dependents if the
parent databases differ.

Release note: None

fqazi committed:
schemachanger: update logic tests for default enable

Previously, the declarative schema changer was disabled
by default. We are enabling it by default, so some logic
test updates are needed. This patch will update logictest
expected files.

Release note: None

more rewriting...
<pkg>: <short description - lowercase, no final period>

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release note (<category, see below>): <what> <show> <why>

disable
<pkg>: <short description - lowercase, no final period>

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release note (<category, see below>): <what> <show> <why>

fqazi committed:
schemachanger: fix ALTER TABLE errors for pg_catalog tables

Previously, the declarative schema changer did not return
the correct error when attempting to modify pg_catalog
tables. To address this, this patch will return the correct
error when trying to modify pg_catalog tables.

Release note: None

fqazi committed:
schemachanger: disable declarative schema changer certain tests

Previously, the declarative schema changer was enabled
by default, which could cause certain tests designed
for the legacy schema changer to fail. To address this,
this patch individual disables declarative schema changer
on tests designed for legacy schema changer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76610)
<!-- Reviewable:end -->

fqazi committed:
schemachanger: clean up type references to multiregion enums

Previously, declarative schema changer did not clean up
type references to multiregion enums when dropping tables.
This was inadequate and could lead to validation errors due
to dangling references. To address this, this patch will
clean up multiregion type references in tables.

Release note: None

fqazi committed:
schemachanger: fix drop type dependency messages

Previously, the message generated for drop types that
were not allowed due to dependencies was had a period.
This patch will update the message to match the legacy
schema changer.

Release note: None

fqazi committed:
schemachanger: drop database should handle corrupt descriptors robustly

Previously, the drop database logic inside the declarative schema
changer always assumed that a public schema would exist. However,
due to corruption its possible for his entry to be missing. To
address this, this patch will add extra logic to validate if
the public schema is valid before proceeding to clean it up.

Release note: None

fqazi committed:
schemachanger: fix error returned when dropping multiregion enums

Previously, the declarative schema changer did not generate
the correct messages when attempting to drop/alter multiregion
enums. To address this patch will update the message generated
for multiregion enums to include a hint to indicate that ALTER
DATABASE should be used for such modifications.

Release note: None

fqazi committed:
schemachanger: add support for KV tracing

Previously, the declarative schema changer did not
support KV tracing. This was inadequate because the
legacy schema changer supported this functionality
which is needed for supportability of the product.
To address this, this patch will add support for
KV tracing inside the declarative schema changer.

Release note: None

fqazi committed:
schemachanger: delete descriptors during DROP DATABASE

Previously, when dropping a database descriptor, the
decalrative schema changer never deleted database
descriptors properly. This led to the descriptors being
left behind, instead of fully being cleaned up. To address,
this patch will delete database descriptors during DROP
DATABASE.

Release note: None

fqazi committed:
schemachanger: populate DropTime when dropping tables

Previously, the declarative schema changer did not
populate the drop time when dropping tables. This
could be problematic when looking at crdb_internal tables
and other scenarios that detect if a table is dropped
based on this time. To address this, this patch starts
populating the DropTime inside the declarative schema changer.

Release note: None

merge into drop time pr
<pkg>: <short description - lowercase, no final period>

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release note (<category, see below>): <what> <show> <why>

remove drop time from test
<pkg>: <short description - lowercase, no final period>

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release note (<category, see below>): <what> <show> <why>

fqazi committed:
schemachanger: add telemetry counters for drop statements

Previously, the telemetry metrics were not properly tracked
for drops inside the declarative schema changer. This was
inadequate because we didn't have feature parity with the
legacy schema changer. To address this, this patch adds
support for tracking telemetry metrics inside the declarative
schema changer.

Release note: None.

fqazi committed:
schemachanger: check commands are supported first for alter table

Previously, the declarative schema changer fetched information
about tables in the transaction for ALTER TABLE before checking
if the commands were supported. This was problematic because
it could change timing/retry behaviour in certain cases, since
we would do potentially extra descriptor look ups. To address this,
this patch checks if ALTER TABLE commands are supported first
to avoid intermittent failures caused by extra round trips.

Release note: None

fqazi committed:
schemachanger: disable declarative for retry errors on enums logictest

Previously, there the enums logictest with the declarative schema
changer code paths enabled could intermittently fail with retry
errors. The design of the test is such that we execute SELECT
queries and schema changes in the same transaction, so automatic
retries are not possible once the result set has been sent
to the client. To address this, this patch will workaround the
issue by disabling declarative schema changer for this test.

Release note: None

fqazi committed:
schemachanger: bump test timeout to avoid timeouts on TenantLogicTest

Previously, we were right at the timeout boundary
for the TenantLogicTest. With the declarative schema
changer enabled we are just past the 45 minute mark causing
this test to fail. This patch bumps up the test timeout to 50
minutes to allow for breathing room.

Release note: None
